train_unet.py

```diff
@@ -0,0 +1,119 @@
+                                          transformations.NiftiToTorchDimensionsReorderTransformation(),
+                                          trfs.Lambda(lambda x: torch.from_numpy(x)),
+                                          transformations.OneHotEncoding([0, 1, 2, 3]),
+                                          trfs.Lambda(
+                                              lambda x: F.pad(x, [0, 0, 0, 0, 5, 0]) if x.shape[1] % 16 != 0 else x),
+                                          trfs.Lambda(lambda x: F.interpolate(x, size=args.input_size)),
+                                          trfs.Lambda(lambda x: x.float())
+                                          ])
+
+    volumes_paths, masks_paths = read_dataset_json(args.dataset_json)
+    volumes_set = datasets.NiftiFolder(volumes_paths, volumes_transformations)
+    masks_set = datasets.NiftiFolder(masks_paths, masks_transformations)
+    combined_set = datasets.CombinedDataset(volumes_set, masks_set)
+    with open(args.division_json, "r") as division_file:
+        indeces = json.load(division_file)
+    train_set = Subset(combined_set, indeces["train"])
+    valid_set = Subset(combined_set, indeces["valid"])
+
+    train_loader = torch.utils.data.DataLoader(train_set, batch_size=args.batch_size, shuffle=True)
+    valid_loader = torch.utils.data.DataLoader(valid_set, batch_size=args.batch_size, shuffle=True)
+
+    model = UNet3D(1, 3).float()
+    model.to(args.device)
+    criterion = DiceLoss()
+    optimizer = optim.Adam(model.parameters(), args.learning_rate)
+    dice = DiceScore()
+    metrics = {"dice_edema": metric_for_class(dice, 0),
+               "dice_non_enhancing": metric_for_class(dice, 1),
+               "dice_enhancing": metric_for_class(dice, 2),
+               }
+
+    last_model_logger = ModelLogger(os.path.join(args.log_dir, "models"))
+    last_state_dict_logger = StateDictsLogger(os.path.join(args.log_dir, "models"))
+    best_model_logger = BestModelLogger(os.path.join(args.log_dir, "best"))
+    best_state_dict_logger = BestStateDictLogger(os.path.join(args.log_dir, "best"))
+    tensorboard_logger = TensorboardLogger(args.log_dir)
+
+    early_stopping = EarlyStopping(args.patience)
+
+    log_parameters(args.log_dir, args)
+    for epoch in range(args.epochs):
+        train_loss, train_metrics = run_training_epoch(model, train_loader, optimizer, criterion, metrics, args.device)
+        valid_loss, valid_metrics = run_validation_epoch(model, valid_loader, criterion, metrics, args.device)
+        print(f"Epoch: {epoch} "
+              f"Train loss: {train_loss:.4f} "
+              f"Valid loss: {valid_loss:.4f} "
+              f"Valid dice edema: {valid_metrics['dice_edema']:.4f} "
+              f"Valid dice non enhancing: {valid_metrics['dice_non_enhancing']:.4f} "
+              f"Valid dice enhancing: {valid_metrics['dice_enhancing']:.4f} ", flush=True)
+
+        last_model_logger.log(model, 0)
```

Comment:
Why always 0th epoch?

Reply:
This is sketchy, I want to cut the number of artifacts produced and I saved it always with the index 0. I changed the implementation to take suffix rather than epoch number.

Reply:
Fixed